[safety-rules] latency optimization

zekun000 opened the pull request:
<!--
Thank you for sending a PR. We appreciate you spending time to help improve the Diem project.

The project is undergoing daily changes. Pull Requests will be reviewed and responded to as time permits.
-->

## Motivation

1. return signature instead of full block
2. avoid updating preferred round to save one call to vault (not related to safety)

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

(Write your answer here.)

## Test Plan

(Share your test plan here. If you changed code, please provide us with clear instructions for verifying that your changes work.)

## Related PRs

(If this PR adds or changes functionality, please take some time to update or suggest changes to the docs at https://developers.diem.com, and link to your PR here.)

## If targeting a release branch, please fill the below out as well

 * Justification and breaking nature (who does it affect? validators, full nodes, tooling, operators, AOS, etc.)
 * Comprehensive test results that demonstrate the fix working and not breaking existing workflows.
 * Why we must have it for V1 launch.
 * What workarounds and alternative we have if we do not push the PR.

zekun000 commented:
all up : 1035 TPS, 4392 ms latency, 5000 ms p99 latency, no expired txns
vs
all up : 962 TPS, 4723 ms latency, 5350 ms p99 latency, backup: -1 Bps, no expired txns, time spent 250 secs

aching commented:
This is awesome, I'll try to take a look later today or tomorrow.  Is there also a latency gain on the low TPS test case?

zekun000 commented:
/canary

bors-libra commented:
:broken_heart: Test Failed - [ci-test](https://github.com/diem/diem/actions/runs/968983342)

zekun000 commented:
fixed tps 10 : 10 TPS, 704 ms latency, 800 ms p99 latency, no expired txns
vs
fixed tps 10 : 10 TPS, 916 ms latency, 1000 ms p99 latency, backup: -1 Bps, no expired txns

~200ms saving

zihaoccc commented:
> fixed tps 10 : 10 TPS, 704 ms latency, 800 ms p99 latency, no expired txns
> vs
> fixed tps 10 : 10 TPS, 916 ms latency, 1000 ms p99 latency, backup: -1 Bps, no expired txns
> 
> ~200ms saving

This is awesome

davidiw reviewed (commented):
this looks like a subset of the broader plan to only pass in the relevant bits of the proposal -- e.g., a hash value to replace the transactions

and then we also get the cool cows

did you break down on what got the wins?

zekun000 commented:
@davidiw I can explore that as well but that breaks the hash compatibility.

yes, avoid sending full block saves ~10-20ms, avoiding the vault call saves another ~20ms, so the signed phase dropped from ~90ms to ~50ms

davidiw commented on consensus/safety-rules/src/tests/suite.rs:
```diff
@@ -565,47 +560,14 @@ fn test_sign_proposal_with_invalid_qc(safety_rules: &Callback) {
     );
 }
 
-fn test_sign_proposal_with_early_preferred_round(safety_rules: &Callback) {
```
why?

davidiw commented on consensus/safety-rules/src/safety_rules.rs:
```diff
@@ -403,14 +403,9 @@ impl SafetyRules {
         }
 
         self.verify_qc(block_data.quorum_cert())?;
-        if self.verify_and_update_preferred_round(block_data.quorum_cert(), &mut safety_data)? {
-            self.persistent_storage.set_safety_data(safety_data)?;
-        }
```
oh I see, you have to read this carefully

davidiw replied:
I guess I would actually argue that signing anything isn't really that great. I'm happy to agree with @JoshLind who will most certainly take my side, maybe :P ... though recent track records suggest we are in for a healthy debate.

zekun000 replied:
I don't disagree that signing anything isn't great (otherwise I could just pass a hash value here), I don't really mind checking the preferred round either if that makes you feel better as long as we don't persist it

davidiw commented on consensus/safety-rules/src/tests/suite.rs:
```diff
@@ -565,47 +560,14 @@ fn test_sign_proposal_with_invalid_qc(safety_rules: &Callback) {
-    let err = safety_rules
-        .sign_proposal(a5.block().block_data().clone())
-        .unwrap_err();
-    assert_eq!(err, Error::IncorrectPreferredRound(0, 2));
```
seems like this should still fail, no? since the parent block is old, I wish we had better documented this test case...

zekun000 replied:
sign_proposal doesn't matter, the proposer is free to sign any garbage, the safety critical piece is upon voting.

davidiw reviewed (commented):
why not more commits in the PR, one commit for all is haaard for these old eyes
